mava/callbacks/hook_mixin.py

```diff
@@ -0,0 +1,195 @@
+# python3
+# Copyright 2021 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Abstract mixin class used to call system component hooks."""
+
+from abc import ABC
+from typing import List
+
+
+class CallbackHookMixin(ABC):
+
+    ######################
+    # system builder hooks
+    ######################
+
+    callbacks: List
+
+    # INITIALISATION
+    def on_building_init_start(self) -> None:
```

Comment:
There is a very large number of hooks. Do we handle this as we may at some point need to inject code at any of these points and in that regard, we can add or remove hooks as necessary?

Reply:
Yes, this is exactly right, so in the future, we might add more hooks if needed, but for now, we are creating the ones we think are most necessary.